presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RewriteExpressionWithConstantVariable.java

```diff
@@ -0,0 +1,462 @@
+import com.facebook.presto.sql.planner.plan.OffsetNode;
+import com.facebook.presto.sql.planner.plan.SampleNode;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+import com.facebook.presto.sql.planner.plan.SortNode;
+import com.facebook.presto.sql.planner.plan.UnnestNode;
+import com.facebook.presto.sql.planner.plan.WindowNode;
+import com.facebook.presto.sql.relational.FunctionResolution;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.SystemSessionProperties.isRewriteExpressionWithConstantEnabled;
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.common.type.DateType.DATE;
+import static com.facebook.presto.common.type.IntegerType.INTEGER;
+import static com.facebook.presto.common.type.VarcharType.VARCHAR;
+import static com.facebook.presto.expressions.LogicalRowExpressions.extractConjuncts;
+import static com.facebook.presto.spi.plan.AggregationNode.singleGroupingSet;
+import static com.facebook.presto.sql.planner.plan.ChildReplacer.replaceChildren;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+
+/**
+ * Get constant from filter and project node. Rewrite expressions in parent nodes with the constant.
+ * For example, for query "select orderkey, orderpriority, avg(totalprice) from orders where orderpriority='3-MEDIUM' group by 1, 2", the query plan changes from
+ * <pre>
+ *     - OutputNode
+ *          orderkey, orderpriority, avg
+ *          - Aggregate
+ *              avg := avg(totalprice)
+ *              Grouping Keys := [orderkey, orderpriority]
+ *              - Filter
+ *                  orderpriority = '3-MEDIUM'
+ *                  - TableScan
+ *                      orderkey := orderkey
+ *                      orderpriority := orderpriority
+ *                      totalprice := totalprice
+ * </pre>
+ * to
+ * <pre>
+ *      - OutputNode
+ *          orderkey, expr_12, avg
+ *          - project
+ *              orderkey := orderkey
+ *              expr_12 := '3-MEDIUM'
+ *              avg := avg
+ *              - Aggregate
+ *                  avg := avg(totalprice)
+ *                  Grouping Keys := [orderkey]
+ *                  - Filter
+ *                      orderpriority = '3-MEDIUM'
+ *                      - TableScan
+ *                          orderkey := orderkey
+ *                          orderpriority := orderpriority
+ *                          totalprice := totalprice
+ * </pre>
+ */
+public class RewriteExpressionWithConstantVariable
+        implements PlanOptimizer
```

Comment:
Is there a reason you didn't make this an iterative optimizer?

Reply:
It needs to visit the query plan bottom up, collect and pass the constant constraint. And it works for different kinds of nodes, while iterative optimizer needs to specify the type of the node to work on.